### ISSUE TITLE
Should I use the metrics output after processing by train.py or those after processing by val.py?

### ISSUE THREAD

**Issue by 01XiaoTian** [ultralytics/ultralytics#17838]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

**Question 1**: I have two improved models and I need to compare their effects. Should I use the metrics output after processing by train.py or those after processing by val.py?

**Question 2**: Are the performance metrics output in the terminal at the end of training and validation consistent with the performance metrics on the pictures in the weights and val folders?

Here is an example. It's the result after performing val.py on the trained weights. They have significantly different data.

![image.png](https://tc-cdn.flowus.cn/oss/41582b6b-3962-4da5-9d11-62e9803a97a6/image.png?time=1732719600&token=8ff47dbed41e3b3f546bbb4511022f41feb44d9f00fce7455ab79cabe41a32b6&role=free)

![image.png](https://tc-cdn.flowus.cn/oss/4d78e1e0-05b5-48a0-a26a-88f238103ed7/image.png?time=1732719600&token=0859efdaaeaf1b73056ace847add0803d03468cc83741582276d64008f795f48&role=free)

thanks!

**Comment by Y-T-G** [comment]
Are you using best.pt? What's the used code for validation?

**Comment by Y-T-G** [comment]
1. Both should be the same.
2. The terminal shows the recall and precision at the confidence threshold that produces the best F1-score. The plots show the best recall and best precision which is at the extreme confidence thresholds.

**Comment by 01XiaoTian** [comment]
> 1. Both should be the same.两者应该是相同的。
> 2. The terminal shows the recall and precision at the confidence threshold that produces the best F1-score. The plots show the best recall and best precision which is at the extreme confidence thresholds.终端在产生最佳 F1 分数的置信度阈值处显示召回率和精度。这些图显示了处于极端置信度阈值的最佳召回率和最佳精确率。

Thank you very much for your answer!
I've figured out the second question, and for the first question, if I need to write a paper, should I use the performance metrics after train(train.py), or the performance metrics after validation（val.py）,Their performance metrics have different numerical values

**Comment by 01XiaoTian** [comment]
> 1. Both should be the same.
> 2. The terminal shows the recall and precision at the confidence threshold that produces the best F1-score. The plots show the best recall and best precision which is at the extreme confidence thresholds.

Thank you very much for your answer!
I've figured out the second question, and for the first question, if I need to write a paper, should I use the performance metrics after train(train.py), or the performance metrics after validation（val.py）,Their performance metrics have different numerical values

**Comment by glenn-jocher** [comment]
@01XiaoTian when writing a paper, it is recommended to use the performance metrics from `val.py`. These metrics are generated under a consistent evaluation framework and provide a standardized baseline for comparing models. Metrics from `train.py` are more reflective of training progress rather than final performance.

**Comment by 01XiaoTian** [comment]
> @01XiaoTian when writing a paper, it is recommended to use the performance metrics from `val.py`. These metrics are generated under a consistent evaluation framework and provide a standardized baseline for comparing models. Metrics from `train.py` are more reflective of training progress rather than final performance.

Thank you so much, I see!

I also understood the difference between the indicators output by the terminal and the indicators in the picture. 

Another question is whether the data 'P' and 'R' at the time of writing in the paper are usually based on the content **record of the terminal output** ( that record the recall and precision at the confidence threshold that produces the best F1-score) or **the data record of the picture** in the file ( that record the best recall and best precision which is at the extreme confidence thresholds.)
For example: this record.Thanks again！
![image](https://github.com/user-attachments/assets/1acb5d77-6761-4813-beba-2371164d6d90)

**Comment by glenn-jocher** [comment]
@01XiaoTian when writing a paper, it is generally advisable to use the metrics recorded in the **terminal output**, as they represent recall and precision values at the confidence threshold that achieves the best F1-score. This provides a balanced and widely accepted measure of your model's performance for evaluation and comparison purposes. Metrics in the plots often reflect extreme cases and are less commonly used for such reporting.

**Comment by L31nard0** [comment]
Why are the validation metrics for train.py and val.py different?

**Comment by L31nard0** [comment]
> > Why are the validation metrics for train.py and val.py different?
> 
> Because we used the validation set data to make corrections (modify) to the training results with the val.py, there are some changes to this performance metric

Thank you

**Comment by glenn-jocher** [comment]
@L31nard0 you're correct. The difference arises because `train.py` evaluates the model during training, often influenced by augmentations or temporary states, while `val.py` performs a post-training evaluation, providing a consistent and final assessment of the model's performance.
